casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala

```diff
@@ -1,96 +1,89 @@
 package coop.rchain.casper.util.rholang
 
+import cats.effect.Sync
 import cats.implicits._
 import com.google.protobuf.ByteString
 import coop.rchain.casper.protocol._
 import coop.rchain.casper.util.ProtoUtil
 import coop.rchain.casper.util.rholang.RuntimeManager.StateHash
-import coop.rchain.catscontrib.TaskContrib._
-import coop.rchain.crypto.codec.Base16
 import coop.rchain.crypto.hash.Blake2b512Random
 import coop.rchain.models.Expr.ExprInstance.GString
 import coop.rchain.models._
 import coop.rchain.rholang.interpreter.accounting.{Cost, CostAccount}
 import coop.rchain.rholang.interpreter.storage.StoragePrinter
 import coop.rchain.rholang.interpreter.{accounting, ChargingReducer, ErrorLog, Runtime}
-import coop.rchain.rspace.internal.{Datum, WaitingContinuation}
-import coop.rchain.rspace.trace.Produce
+import coop.rchain.rspace.internal.Datum
 import coop.rchain.rspace.{Blake2b256Hash, ReplayException}
-import monix.eval.Task
-import monix.execution.Scheduler
 
-import coop.rchain.catscontrib.TaskContrib._
-
-import scala.annotation.tailrec
 import scala.collection.immutable
 import scala.concurrent.SyncVar
-import scala.concurrent.duration._
 import scala.util.{Failure, Success, Try}
+import scala.language.higherKinds
 
 //runtime is a SyncVar for thread-safety, as all checkpoints share the same "hot store"
-class RuntimeManager private (val emptyStateHash: ByteString, runtimeContainer: SyncVar[Runtime]) {
-
-  def captureResults(start: StateHash, deploy: Deploy, name: String = "__SCALA__")(
-      implicit scheduler: Scheduler
-  ): Seq[Par] = {
-    val runtime                   = runtimeContainer.take()
-    val (_, Seq(processedDeploy)) = newEval(deploy :: Nil, runtime, start).unsafeRunSync
-
-    //TODO: Is better error handling needed here?
-    val result: Seq[Datum[ListParWithRandom]] =
-      if (processedDeploy.status.isFailed) Nil
-      else {
-        val returnChannel = Par().copy(exprs = Seq(Expr(GString(name))))
-        runtime.space.getData(returnChannel).unsafeRunSync
-      }
+class RuntimeManager[F[_]: Sync] private (
+    val emptyStateHash: ByteString,
+    runtimeContainer: SyncVar[Runtime[F]]
+) {
 
-    runtimeContainer.put(runtime)
+  def captureResults(start: StateHash, deploy: Deploy, name: String = "__SCALA__"): F[Seq[Par]] =
+    for {
+      runtime    <- Sync[F].delay { runtimeContainer.take() }
+      evalResult <- newEval(deploy :: Nil, runtime, start)
 
-    result.flatMap(_.a.pars)
-  }
+      (_, Seq(processedDeploy)) = evalResult
+
+      //TODO: Is better error handling needed here?
+//      _ <- if (processedDeploy.status.isFailed)
+//            Sync[F].raiseError(new Exception("Processed deploy has failed"))
+//          else ()
+
+      returnChannel = Par().copy(exprs = Seq(Expr(GString(name))))
+      result        <- runtime.space.getData(returnChannel)
+
+      _ <- Sync[F].delay { runtimeContainer.put(runtime) }
+    } yield result.flatMap(_.a.pars)
```

Comment:
So what happens if the call to captureResults fails? Does it throw a stacktrace and crash the node?

Reply:
Thje code was commented by mistake. It's fixed now